File: apps/web/src/api/changeReview.ts
import { get, post } from './client';

export const changeStatus = (status: string) => ({ proposed: '待审批', approved: '已批准，待部署', rejected: '已驳回', deploying: '部署中', effective: '后端标记已生效', failed: '部署失败', rolled_back: '已回滚', emergency_applied: '紧急已批准，仍需核对执行与复核', post_review_due: '待复核' }[status] ?? '状态待核对');
export const modeLabel = (mode: string) => ({ audit_only: '仅记录', warn: '提醒', block: '拦截' }[mode] ?? '未知模式');
export const blockerLabels: Record<string, string> = {
  missing_permission: '当前账号没有审批权限，请联系组织管理员。', not_proposed: '此变更已处理，不能再次审批。',
  own_proposal: '提出者不能批准自己的变更，请由另一位有权限的审批人处理。',
  incomplete_content: '内容包含已隐藏或超出展示上限的部分，无法完整审查；请提出者整理后重新提交。',
  policy_invalid: '策略尚未通过静态校验，请提出者修正后重新提交。',
  downgrade_requires_high_risk: '保护模式低于前一版本，需要按高风险变更重新提交。',
};
export interface ChangeReview {
  schema_version: 'change-review/v1'; change_id: string; status: string; policy_name: string; policy_version: number;
  enforcement_mode: 'audit_only' | 'warn' | 'block'; approval_policy: 'standard' | 'high_risk' | 'break_glass';
  proposer_id: string; approver_id: string | null; review_digest: string;
  sections: { key: string; label: string; content: string; redacted: boolean; truncated: boolean }[];
  can_approve: boolean; can_reject: boolean; approve_blockers: string[]; reject_blockers: string[];
}
const object = (v: unknown): v is Record<string, unknown> => !!v && typeof v === 'object' && !Array.isArray(v);
const text = (v: unknown, max = 256): v is string => typeof v === 'string' && v.length <= max;
const keys = (v: Record<string, unknown>, expected: string[]) => Object.keys(v).length === expected.length && expected.every(k => k in v);
export function parseChangeReview(value: unknown, id: string): ChangeReview {
  if (!object(value) || !keys(value, ['schema_version', 'change_id', 'status', 'policy_name', 'policy_version', 'enforcement_mode', 'approval_policy', 'proposer_id', 'approver_id', 'review_digest', 'sections', 'can_approve', 'can_reject', 'approve_blockers', 'reject_blockers']) || value.schema_version !== 'change-review/v1' || value.change_id !== id || !text(value.status, 64) || !text(value.policy_name, 6000) || !Number.isInteger(value.policy_version) || (value.policy_version as number) < 1 || !['audit_only', 'warn', 'block'].includes(String(value.enforcement_mode)) || !['standard', 'high_risk', 'break_glass'].includes(String(value.approval_policy)) || !text(value.proposer_id) || !(value.approver_id === null || text(value.approver_id)) || !text(value.review_digest) || !/^[a-f0-9]{64}$/.test(value.review_digest) || typeof value.can_approve !== 'boolean' || typeof value.can_reject !== 'boolean') throw new Error('审查响应不完整或对象不匹配');
  for (const k of ['approve_blockers', 'reject_blockers'] as const) {
    const list = value[k];
    if (!Array.isArray(list) || list.length > (k === 'approve_blockers' ? 6 : 2) || !list.every(x => typeof x === 'string' && Object.hasOwn(blockerLabels, x)) || new Set(list).size !== list.length) throw new Error('审批权限响应不完整');
  }
  if (value.can_approve !== ((value.approve_blockers as string[]).length === 0) || value.can_reject !== ((value.reject_blockers as string[]).length === 0)) throw new Error('审批权限响应不一致');
  if (!Array.isArray(value.sections) || value.sections.length !== 15 || !value.sections.every(s => object(s) && keys(s, ['key', 'label', 'content', 'redacted', 'truncated']) && text(s.key, 64) && text(s.label, 128) && text(s.content, 6000) && typeof s.redacted === 'boolean' && typeof s.truncated === 'boolean')) throw new Error('策略内容响应不完整');
  const required = ['selector', 'filesystem', 'network', 'process', 'model_routing', 'tools', 'tool_policies', 'data_scope_refs', 'secrets', 'resources', 'audit', 'exceptions', 'impact', 'validation', 'previous'];
  if (!required.every(k => (value.sections as ChangeReview['sections']).some(s => s.key === k)) || value.can_approve && (value.status !== 'proposed' || value.sections.some(s => s.redacted || s.truncated))) throw new Error('审查章节缺失或审批响应不一致');
  return value as unknown as ChangeReview;
}
export async function readChangeReview(id: string): Promise<ChangeReview> {
  return parseChangeReview(await get<unknown>(`/change-requests/${encodeURIComponent(id)}/review`), id);
}
export function submitChangeReview(snapshot: ChangeReview, decision: 'approve' | 'reject'): Promise<unknown> {
  return post(`/change-requests/${encodeURIComponent(snapshot.change_id)}/review-decision`, { schema_version: 'change-review-decision/v1', decision, review_digest: snapshot.review_digest });
}

File: apps/control-api/app/routers/change_review.py
"""Bounded, identity-bound review snapshots; legacy decision routes remain compatible."""

from __future__ import annotations

import hashlib
import hmac
import json
import re
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_session
from app.models import ChangeRequest, DesiredPolicy
from app.routers.policies import (
    _MODE_RANK,
    _approve_change_request,
    _desired_from_policy,
    _reject_change_request,
    _validate_policy_static,
)
from app.rulepack import load_rulepack
from app.schemas import ChangeRequestOut
from app.security import Identity, ensure_permission, get_identity

router = APIRouter(tags=["policies"])
_SENSITIVE_KEY = re.compile(
    r"^(?:password|passwd|(?:access[_-]?|refresh[_-]?|id[_-]?)?token|credentials?|authorization|"
    r"api[_-]?key|private[_-]?key|signing[_-]?key|client[_-]?secret|secret(?:[_-]?(?:value|key))?)$",
    re.I,
)
_DOMAINS = {
    "selector": "适用对象（技术标识）",
    "filesystem": "文件访问",
    "network": "网络访问",
    "process": "进程约束",
    "model_routing": "模型路由",
    "tools": "工具清单",
    "tool_policies": "工具权限",
    "data_scope_refs": "数据范围引用",
    "secrets": "凭据引用",
    "resources": "资源限制",
    "audit": "审计要求",
    "exceptions": "例外规则",
}


class Wire(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Section(Wire):
    key: str = Field(max_length=64)
    label: str = Field(max_length=128)
    content: str = Field(max_length=6000)
    redacted: bool
    truncated: bool


ApproveBlocker = Literal[
    "missing_permission",
    "not_proposed",
    "own_proposal",
    "incomplete_content",
    "policy_invalid",
    "downgrade_requires_high_risk",
]
RejectBlocker = Literal["missing_permission", "not_proposed"]


class ChangeReview(Wire):
    schema_version: Literal["change-review/v1"] = "change-review/v1"
    change_id: str = Field(max_length=256)
    status: str = Field(max_length=64)
    policy_name: str = Field(max_length=6000)
    policy_version: int = Field(ge=1)
    enforcement_mode: Literal["audit_only", "warn", "block"]
    approval_policy: Literal["standard", "high_risk", "break_glass"]
    proposer_id: str = Field(max_length=256)
    approver_id: str | None = Field(max_length=256)
    review_digest: str = Field(pattern=r"^[a-f0-9]{64}$")
    sections: list[Section] = Field(max_length=16)
    can_approve: bool
    can_reject: bool
    approve_blockers: list[ApproveBlocker] = Field(max_length=6)
    reject_blockers: list[RejectBlocker] = Field(max_length=2)


class ReviewDecision(Wire):
    schema_version: Literal["change-review-decision/v1"]
    decision: Literal["approve", "reject"]
    review_digest: str = Field(pattern=r"^[a-f0-9]{64}$")


def _section(key: str, label: str, value: object, rules: tuple) -> Section:
    redacted = False
    truncated = False
    nodes = 0

    def clean(item: object, depth: int = 0) -> object:
        nonlocal redacted, truncated, nodes
        nodes += 1
        if nodes > 1000 or depth > 16:
            truncated = True
            return "[内容超出审查上限]"
        if isinstance(item, dict):
            result = {}
            for index, (k, v) in enumerate(item.items()):
                if index >= 1000 or nodes > 1000:
                    truncated = True
                    break
                safe_key = str(clean(str(k), depth + 1))
                if _SENSITIVE_KEY.search(str(k)):
                    redacted = True
                    result[safe_key] = "[已隐藏敏感字段]"
                else:
                    result[safe_key] = clean(v, depth + 1)
            return result
        if isinstance(item, list):
            result = []
            for v in item:
                if nodes > 1000:
                    truncated = True
                    break
                result.append(clean(v, depth + 1))
            return result
        if isinstance(item, str):
            # Do not return a long prefix that might split an unrecognised secret.
            if len(item) > 6000:
                truncated = True
                return "[文本超出审查上限]"
            safe = item
            for rule in rules:
                safe = rule.pattern.sub(rule.replacement, safe)
            redacted |= safe != item
            return safe
        return item

    safe = clean(value)
    content = (
        json.dumps(safe, ensure_ascii=False, indent=2) if value is not None else "未配置（具体默认行为由执行后端决定）"
    )
    if len(content) > 6000:
        content = "内容超出审查上限，请拆分策略后重新提交。"
        truncated = True
    return Section(key=key, label=label, content=content, redacted=redacted, truncated=truncated)


def _locate(session: Session, identity: Identity, cr_id: str, *, lock: bool = False):
    query = select(ChangeRequest).where(ChangeRequest.id == cr_id, ChangeRequest.tenant_id == identity.tenant_id)
    cr = session.scalar(query.with_for_update() if lock else query)
    if cr is None:
        raise HTTPException(404, "not_found")
    ensure_permission(identity, "policy:read")
    query = select(DesiredPolicy).where(DesiredPolicy.id == cr.policy_id, DesiredPolicy.tenant_id == identity.tenant_id)
    policy = session.scalar(query.with_for_update() if lock else query)
    if policy is None:
        raise HTTPException(404, "not_found")
    return cr, policy


def _snapshot(session: Session, identity: Identity, cr: ChangeRequest, policy: DesiredPolicy) -> ChangeReview:
    previous = session.scalar(
        select(DesiredPolicy)
        .where(
            DesiredPolicy.tenant_id == identity.tenant_id,
            DesiredPolicy.name == policy.name,
            DesiredPolicy.version < policy.version,
        )
        .order_by(DesiredPolicy.version.desc(), DesiredPolicy.id)
    )
    previous_basis = (
        {"id": previous.id, "mode": previous.enforcement_mode, "status": previous.status} if previous else None
    )
    desired = _desired_from_policy(policy)
    _, _, rules = load_rulepack()
    title = _section("name", "策略名称", policy.name, rules)
    sections = [_section(key, label, desired[key], rules) for key, label in _DOMAINS.items()]
    sections += [
        _section("impact", "申请人填写的用途与影响（尚未经系统核验）", cr.impact, rules),
        _section("validation", "校验提示", policy.unsupported_by_backend, rules),
        _section("previous", "前一版本模式（不代表当前运行时状态）", previous_basis, rules),
    ]
    reject: list[RejectBlocker] = []
    if not identity.has_permission("change:approve"):
        reject.append("missing_permission")
    if cr.status != "proposed":
        reject.append("not_proposed")
    approve: list[ApproveBlocker] = list(reject)
    if cr.proposer_user_id == identity.actor_id:
        approve.append("own_proposal")
    if any(s.redacted or s.truncated for s in [title, *sections]):
        approve.append("incomplete_content")
    if policy.status != "validated" or _validate_policy_static(policy):
        approve.append("policy_invalid")
    if (
        previous
        and previous.status not in ("rejected", "failed")
        and _MODE_RANK[policy.enforcement_mode] < _MODE_RANK.get(previous.enforcement_mode, -1)
        and cr.approval_policy != "high_risk"
    ):
        approve.append("downgrade_requires_high_risk")
    digest = hashlib.sha256(
        json.dumps(
            {
                "version": "change-review/v1",
                "tenant": identity.tenant_id,
                "actor": identity.actor_id,
                "actor_type": identity.identity_type,
                "change": cr.id,
                "status": cr.status,
                "proposer": cr.proposer_user_id,
                "approver": cr.approver_user_id,
                "approval_policy": cr.approval_policy,
                "impact": cr.impact,
                "diff": cr.diff,
                "policy": desired,
                "name": policy.name,
                "validation": policy.unsupported_by_backend,
                "previous": previous_basis,
            },
            ensure_ascii=False,
            sort_keys=True,
            separators=(",", ":"),
        ).encode()
    ).hexdigest()
    return ChangeReview(
        change_id=cr.id,
        status=cr.status,
        policy_name=json.loads(title.content) if not title.truncated else "名称过长",
        policy_version=policy.version,
        enforcement_mode=policy.enforcement_mode,
        approval_policy=cr.approval_policy,
        proposer_id=cr.proposer_user_id,
        approver_id=cr.approver_user_id,
        review_digest=digest,
        sections=sections,
        can_approve=not approve,
        can_reject=not reject,
        approve_blockers=approve,
        reject_blockers=reject,
    )


@router.get("/api/v1/change-requests/{cr_id}/review", response_model=ChangeReview)
def get_review(
    cr_id: str, response: Response, session: Session = Depends(get_session), identity: Identity = Depends(get_identity)
):
    cr, policy = _locate(session, identity, cr_id)
    response.headers["Cache-Control"] = "no-store"
    return _snapshot(session, identity, cr, policy)


@router.post("/api/v1/change-requests/{cr_id}/review-decision", response_model=ChangeRequestOut)
def review_decision(
    cr_id: str,
    body: ReviewDecision,
    session: Session = Depends(get_session),
    identity: Identity = Depends(get_identity),
):
    cr, policy = _locate(session, identity, cr_id, lock=True)
    ensure_permission(identity, "change:approve")
    current = _snapshot(session, identity, cr, policy)
    if not hmac.compare_digest(body.review_digest, current.review_digest):
        raise HTTPException(409, "review_changed")
    blockers = current.approve_blockers if body.decision == "approve" else current.reject_blockers
    if blockers:
        raise HTTPException(409, f"review_blocked:{blockers[0]}")
    apply = _approve_change_request if body.decision == "approve" else _reject_change_request
    return apply(cr, session, identity, review_digest=current.review_digest)
